src/internal/components/file-token-group/file-token.tsx

```diff
@@ -0,0 +1,164 @@
+import InternalSpinner from '../../../spinner/internal.js';
+import { TokenGroupProps } from '../../../token-group/interfaces.js';
+import { Token } from '../../../token-group/token.js';
+import { BaseComponentProps } from '../../base-component/index.js';
+import Tooltip from '../tooltip/index';
+import * as defaultFormatters from './default-formatters.js';
+import { FileOptionThumbnail } from './thumbnail.js';
+
+import styles from './styles.css.js';
+import testUtilStyles from './test-classes/styles.css.js';
+
+export namespace FileTokenProps {
+  export interface I18nStrings {
+    removeFileAriaLabel: (fileIndex: number) => string;
+    errorIconAriaLabel?: string;
+    warningIconAriaLabel?: string;
+    formatFileSize?: (sizeInBytes: number) => string;
+    formatFileLastModified?: (date: Date) => string;
+  }
+}
+
+export interface FileTokenProps extends BaseComponentProps {
+  file: File;
+  onDismiss: () => void;
+  showFileSize?: boolean;
+  showFileLastModified?: boolean;
+  showFileThumbnail?: boolean;
+  errorText?: React.ReactNode;
+  warningText?: React.ReactNode;
+  loading?: boolean;
+  loadingText?: string;
+  i18nStrings: FileTokenProps.I18nStrings;
+  dismissLabel?: string;
+  alignment?: TokenGroupProps.Alignment;
+  groupContainsImage?: boolean;
+  index: number;
+}
+
+function InternalFileToken({
+  file,
+  showFileLastModified,
+  showFileSize,
+  showFileThumbnail,
+  i18nStrings,
+  onDismiss,
+  errorText,
+  warningText,
+  loading,
+  loadingText,
+  alignment,
+  groupContainsImage,
+  index,
+}: FileTokenProps) {
+  const isImage = file.type.startsWith('image/');
+  const formatFileSize = i18nStrings.formatFileSize ?? defaultFormatters.formatFileSize;
+  const formatFileLastModified = i18nStrings.formatFileLastModified ?? defaultFormatters.formatFileLastModified;
+
+  const containerRef = useRef<HTMLDivElement>(null);
+  const fileNameRef = useRef<HTMLSpanElement>(null);
+  const fileNameContainerRef = useRef<HTMLDivElement>(null);
+  const [showTooltip, setShowTooltip] = useState(false);
+
+  function isEllipsisActive() {
+    const span = fileNameRef.current;
+    const container = fileNameContainerRef.current;
+
+    if (span && container) {
+      return span.offsetWidth >= container.offsetWidth;
+    }
+    return false;
+  }
+
+  return (
+    <div ref={containerRef} className={clsx(styles['file-token'])}>
+      <Token
+        type="file"
+        ariaLabel={file.name}
+        dismissLabel={i18nStrings.removeFileAriaLabel(index)}
+        onDismiss={onDismiss}
+        errorText={errorText}
+        warningText={warningText}
+        errorIconAriaLabel={i18nStrings.errorIconAriaLabel}
+        warningIconAriaLabel={i18nStrings.warningIconAriaLabel}
+        loading={loading}
+        alignment={alignment}
+        groupContainsImage={groupContainsImage && showFileThumbnail && alignment === 'horizontal'}
+        data-index={index}
+      >
+        {loading && (
+          <>
```

Comment:
Updated: I found the loading text at the bottom. Would it make sense to render it here instead?

Reply:
I wonder of the loading text is a good approach for the file upload though. In theory, there can be a large number of files uploading at the same time, making the announcements potentially annoying and not legible. This can be solved by not providing the loading text and handle it from the outside, but maybe a nicer approach could be to defined some getLoadingText function that takes files as input and generates a single loading text for all files like "3 files uploading, 79% complete".

Reply:
Removed this property - discussed adding a guideline to add a separate `LiveRegion`.